### PR TITLE
Benchmarking also benchmark for decoding the call

### PR DESCRIPTION
**Pull request by thiolliere**
in the benchmarking macro, when user use the syntax: `bench_name { $code }: $dispatch_name($origin, $args) verify { $code }`, the benchmark only dispatch the call. This PR changes it, instead the benchmark decode the call and dispatch it.

## Breaking change

Some benchmark code won't compile complaining about partially moved variables. It should be generally easily fixable using clone.

### CODE REVIEW

**Comment by kianenigma** [comment]
Sensible to have it IMO 👍🏻

**Comment by shawntabrizi** [comment]
I agree, if we are finding this is more and more relevant, we should definitely include it.

How can we add a test so that we know this is working as expected?

**Comment by emostov** [comment]
> I agree, if we are finding this is more and more relevant, we should definitely include it.
> 
> How can we add a test so that we know this is working as expected?

Was there any specific rational for not including it before?

From my limited understanding it makes sense to always do this iff when processing the extrinsics in a block we always have to decode each one.

**Review by emostov — APPROVED** [review]
My macro literacy is not great, but from what I understand this makes sense. 

Should we re-benchmark everything (including in polkadot) so when we make logic changes in the future they are based off a baseline that used the same benchmarking formula? Otherwise I would think all future benchmarks would be a bit high.

**Review by shawntabrizi — APPROVED** [review]

**Comment by thiolliere** [comment]
benchmarking in polkadot is generally done before releasing the runtime.
But I'll run benchmark for pallets in substrate

**Comment by thiolliere** [comment]
/benchmark runtime pallet pallet_election_provider_multi_phase

File: frame/system/src/weights.rs [hunk 1]
```diff
@@ -55,79 +55,83 @@ pub trait WeightInfo {
 
 /// Weights for frame_system using the Substrate node and recommended hardware.
 pub struct SubstrateWeight<T>(PhantomData<T>);
-impl<T: crate::Config> WeightInfo for SubstrateWeight<T> {
-	fn remark(_b: u32, ) -> Weight {
-		(1_038_000 as Weight)
+impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
```
Comment (shawntabrizi) [inline frame/system/src/weights.rs:58]:
```suggestion
impl<T: crate::Config> WeightInfo for SubstrateWeight<T> {
```

**Comment by shawntabrizi** [comment]
bot merge

File: frame/system/src/weights.rs [hunk 2]
```diff
@@ -56,78 +56,82 @@ pub trait WeightInfo {
...
-			.saturating_add((795_000 as Weight).saturating_mul(p as Weight))
+			.saturating_add((781_000 as Weight).saturating_mul(p as Weight))
 			.saturating_add(T::DbWeight::get().writes((1 as Weight).saturating_mul(p as Weight)))
 	}
 }
 
 // For backwards compatibility and tests
 impl WeightInfo for () {
-	fn remark(_b: u32, ) -> Weight {
-		(1_038_000 as Weight)
```
Comment (thiolliere) [inline frame/system/src/weights.rs:99]: interesting the weight is now only factor of message length

**Comment by thiolliere** [comment]
polkadot build will be fixed by https://github.com/paritytech/polkadot/pull/3507

**Review by kianenigma — APPROVED** [review]
LGTM

**Comment by thiolliere** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**